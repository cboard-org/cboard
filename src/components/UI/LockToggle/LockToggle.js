import React from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from 'react-intl';
import LockOutlinedIcon from '@material-ui/icons/LockOutlined';
import LockOpenIcon from '@material-ui/icons/LockOpen';

import withChildProof from './childProof';
import IconButton from '../IconButton';
import messages from './LockToggle.messages';

const propTypes = {
  /**
   * @ignore
   */
  intl: intlShape.isRequired,
  /**
   * If true, toggle will be locked
   */
  locked: PropTypes.bool
};

export function LockToggle(props) {
  const { intl, locked, ...rest } = props;

  const lockButtonLabel = locked
    ? intl.formatMessage(messages.unlock)
    : intl.formatMessage(messages.lock);

  return (
    <IconButton label={lockButtonLabel} {...rest}>
      {locked ? <LockOutlinedIcon /> : <LockOpenIcon />}
    </IconButton>
  );
}

LockToggle.propTypes = propTypes;

const ChildProofLockToggle = withChildProof(LockToggle);

export default injectIntl(ChildProofLockToggle);
