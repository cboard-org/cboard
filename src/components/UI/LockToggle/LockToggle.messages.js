import { defineMessages } from 'react-intl';

export default defineMessages({
  lock: {
    id: 'cboard.components.LockToggle.lock',
    defaultMessage: 'Lock'
  },
  unlock: {
    id: 'cboard.components.LockToggle.unlock',
    defaultMessage: 'Unlock'
  }
});
