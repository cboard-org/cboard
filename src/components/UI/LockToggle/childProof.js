import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash/debounce';

function withChildProof(WrappedComponent) {
  return class extends PureComponent {
    static propTypes = {
      /**
       * Clicks to unlock, when number of clicks reached fires onClick
       */
      clicksToUnlock: PropTypes.number,
      /**
       * Callback fired on unlock and lock
       */
      onClick: PropTypes.func.isRequired,
      /**
       * Callback fired on each click
       */
      onLockTick: PropTypes.func
    };

    static defaultProps = {
      clicksToUnlock: 4,
      onLockTick: () => {}
    };

    count = 0;

    incrementCount() {
      this.count = this.count + 1;
    }

    resetCount() {
      this.count = 0;
    }

    debouncedResetCount = debounce(() => {
      this.resetCount();
    }, 2000);

    tickLock(onToggle, onTick) {
      const { clicksToUnlock } = this.props;

      this.incrementCount();

      if (this.count < clicksToUnlock) {
        this.debouncedResetCount();
      } else {
        this.debouncedResetCount.cancel();
      }

      if (this.count <= clicksToUnlock) {
        const clicksLeft = clicksToUnlock - this.count;
        onTick(clicksLeft);
      }

      if (this.count === clicksToUnlock) {
        onToggle();
      }

      if (this.count === clicksToUnlock + 1) {
        this.resetCount();
        onToggle();
      }
    }

    handleClick = event => {
      const { onClick, onLockTick } = this.props;

      this.tickLock(() => {
        onClick(event);
      }, onLockTick);
    };

    render() {
      return <WrappedComponent {...this.props} onClick={this.handleClick} />;
    }
  };
}

export default withChildProof;
