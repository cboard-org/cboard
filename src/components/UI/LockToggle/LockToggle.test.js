import React from 'react';
import { shallowMatchSnapshot } from '../../../common/test_utils';

import LockToggle from './LockToggle';

describe('LockToggle tests', () => {
  test('default renderer', () => {
    shallowMatchSnapshot(<LockToggle onClick={() => {}} />);
  });
});
