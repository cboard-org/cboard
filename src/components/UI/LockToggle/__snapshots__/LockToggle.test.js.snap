// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LockToggle tests default renderer 1`] = `
<_class
  clicksToUnlock={4}
  intl={
    Object {
      "formatMessage": [Function],
    }
  }
  onClick={[Function]}
  onLockTick={[Function]}
/>
`;
